File: tutorial/lect3-helloworld/runSentences_soln.m
try; cd(fileparts(mfilename('fullpath')));catch; end;
run ../../utilities/initPaths.m

buffhost='localhost';buffport=1972;
% wait for the buffer to return valid header information
hdr=[];
while ( isempty(hdr) || ~isstruct(hdr) || (hdr.nchans==0) ) % wait for the buffer to contain valid data
  try 
    hdr=buffer('get_hdr',[],buffhost,buffport); 
  catch
    hdr=[];
    fprintf('Invalid header info... waiting.\n');
  end;
  pause(1);
end;

% set the real-time-clock to use
initgetwTime;
initsleepSec;

% make the target sequence
sentences={'hello world','this is new!','BCI is fun!'};
interSentenceDuration=3;
interCharDuration=1;

% make the stimulus
clf;
set(gcf,'color',[0 0 0],'toolbar','none','menubar','none'); % black figure
set(gca,'visible','off','color',[0 0 0]); % black axes
h=text(.5,.5,'text','HorizontalAlignment','center','VerticalAlignment','middle',...
       'FontUnits','normalized','fontsize',.2,'color',[1 1 1],'visible','off'); 

% play the stimulus
sendEvent('stimulus.sentences','start');
for si=1:numel(sentences);
  sentence=sentences{si};
      
  % reset the cue and fixation point to indicate trial has finished  
  set(h,'visible','off');
  drawnow;
  sendEvent('stimulus.sentence',sentence);
    
  % loop over characters in the sentence
  for ci=1:numel(sentence);
    char = sentence(ci);
    sendEvent('stimulus.character',char);
    set(h,'string',sentence(1:ci),'visible','on');drawnow;
    sleepSec(interCharDuration);
  end
  sleepSec(interSentenceDuration);
  
  % wait for a key press
  msg=msgbox({'Press OK to continue'},'Continue?');while ishandle(msg); pause(.2); end;
  
end % sequences
% end training marker
sendEvent('stimulus.sentences','end');
% tell user we're done
msg=msgbox({'Thanks for taking part!'},'Continue?');while ishandle(msg); pause(.2); end;

File: tutorial/lect3-helloworld/runSentences_soln.py
#!/usr/bin/env python3
# Set up imports and paths
import matplotlib.pyplot as plt
import sys, os
from time import sleep, time
bufhelpPath = "../../python/signalProc"
sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)),bufhelpPath))
import bufhelp

## HELPER FUNCTIONS
def drawnow(fig=None):
    "force a matplotlib figure to redraw itself, inside a compute loop"
    if fig is None: fig=plt.gcf()
    fig.canvas.draw()
    plt.pause(1e-3) # wait for draw.. 1ms

currentKey=None
def keypressFn(event):
    "wait for keypress in a matplotlib figure, and store in the currentKey global"
    global currentKey
    currentKey=event.key
def waitforkey(fig=None,reset=True):
    "wait for a key to be pressed in the given figure"
    if fig is None: fig=gcf()
    global currentKey
    fig.canvas.mpl_connect('key_press_event',keypressFn)
    if reset: currentKey=None
    while currentKey is None:
        plt.pause(1e-2) # allow gui event processing

## CONFIGURABLE VARIABLES

# make the target sequence
sentences=['hello world','this is new!','BCI is fun!'];
interSentenceDuration=3;
interCharDuration=1;
    
##--------------------- Start of the actual experiment loop ----------------------------------
# set the display and the string for stimulus
fig = plt.figure()
fig.suptitle('RunSentences-Stimulus', fontsize=14, fontweight='bold')
ax = fig.add_subplot(111) # default full-screen ax
ax.set_xlim((-1,1))
ax.set_ylim((-1,1))
ax.set_axis_off()
txthdl =ax.text(0, 0, 'This is some text', style='italic')

## init connection to the buffer
ftc,hdr=bufhelp.connect();

bufhelp.sendEvent('stimulus.sentences','start');
## STARTING PROGRAM LOOP
for si,sentence in enumerate(sentences):
    
    # reset the display
    txthdl.set(text='')
    drawnow()
    bufhelp.sendEvent('stimulus.sentence',sentence)

    for ci in range(len(sentence)):
        char = sentence[ci]
        print("%d) char=%s sentence=%s"%(ci,char,sentence[0:ci+1]))
        txthdl.set(text=sentence[0:ci+1])
        drawnow()
        bufhelp.sendEvent('stimulus.character',char)
        sleep(interCharDuration)
    
    drawnow()
    sleep(interSentenceDuration)
    
    #wait for key-press to continue
    txthdl.set(text='Press key to continue')
    drawnow()
    waitforkey(fig)

bufhelp.sendEvent('stimulus.sentences','end')
txthdl.set(text=['Thanks for taking part!' '' 'Press key to finish'])
waitforkey(fig)
